libsolutil/LP.h

```diff
@@ -0,0 +1,88 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libsolutil/Numeric.h>
+
+#include <boost/rational.hpp>
+
+#include <vector>
+#include <variant>
+
+namespace solidity::util
+{
+
+/**
+ * Constraint of the form
+ *  - data[1] * x_1 + data[2] * x_2 + ... <= data[0]  (equality == false)
+ *  - data[1] * x_1 + data[2] * x_2 + ...  = data[0]  (equality == true)
+ * The set and order of variables is implied.
+ */
+struct Constraint
+{
+	std::vector<boost::rational<bigint>> data;
+	bool equality = false;
+
+	bool operator<(Constraint const& _other) const;
+	bool operator==(Constraint const& _other) const;
+};
+
+/**
+ * State used when solving an LP problem.
+ */
+struct SolvingState
+{
+	/// Names of variables, the index zero should be left empty.
+	/// TODO can we change that?
```

Comment:
Use a wrapper class around vector that asserts the read/written index is never 0?